File: src/main/java/com/machfour/macros/core/BaseTable.java
package com.machfour.macros.core;

import java.util.*;

abstract class BaseTable<M> implements Table<M> {
    private final String name;
    private final Factory<M> factory;
    private final List<Column<M, ?>> columns;
    private final List<Column.Fk<M, ?, ?>> fkColumns;
    private final List<Column<M, ?>> secondaryKeyCols;
    private final Map<String, Column<M, ?>> columnsByName;
    private final Column<M, Long> idColumn;
    private final Column<M, Long> createTimeColumn;
    private final Column<M, Long> modifyTimeColumn;
    private final Column<M, ?> naturalKeyColumn;

    // first three columns must be ID, create time, modify time
    BaseTable(String name, Factory<M> factory, List<Column<M, ?>> cols) {
        this.name = name;
        this.factory = factory;

        // TODO make this better
        this.idColumn = (Column<M, Long>)cols.get(0);
        this.createTimeColumn = (Column<M, Long>)cols.get(1);
        this.modifyTimeColumn = (Column<M, Long>)cols.get(2);

        assert Schema.ID_COLUMN_NAME.equals(this.idColumn.sqlName());
        assert Schema.CREATE_TIME_COLUMN_NAME.equals(this.createTimeColumn.sqlName());
        assert Schema.MODIFY_TIME_COLUMN_NAME.equals(this.modifyTimeColumn.sqlName());


        this.columns = Collections.unmodifiableList(cols);

        Column<M, ?> naturalKeyColumn = null;
        // make name map and secondary key cols list. Linked hash map keeps insertion order
        Map<String, Column<M, ?>> columnsByName = new LinkedHashMap<>(cols.size(), 1);
        List<Column<M, ?>> secondaryKeyCols = new ArrayList<>(2);
        List<Column.Fk<M, ?, ?>> fkColumns = new ArrayList<>(2);

        int index = 0;
        for (Column<M, ?> c : cols) {
            setTable(c, index);

            columnsByName.put(c.sqlName(), c);
            if (c.inSecondaryKey()) {
                secondaryKeyCols.add(c);
            }
            // record secondary key
            if (c.isUnique() && !c.equals(this.idColumn)) {
                assert naturalKeyColumn == null : "two natural keys defined";
                naturalKeyColumn = c;
            }
            checkAndAddFk(fkColumns, c);
            index++;
        }
        this.columnsByName = Collections.unmodifiableMap(columnsByName);
        this.secondaryKeyCols = Collections.unmodifiableList(secondaryKeyCols);
        this.fkColumns = Collections.unmodifiableList(fkColumns);
        this.naturalKeyColumn = naturalKeyColumn;
    }

    @SuppressWarnings("unchecked")
    // There's no reason for this check to fail, as the parameter M is guaranteed to match,
    // and everything else is a wildcard
    private static <M> void checkAndAddFk(List<Column.Fk<M, ?, ?>> fkCols, Column<M, ?> c) {
        if (c instanceof Column.Fk) {
            Column.Fk<M, ?, ?> fk = (Column.Fk<M, ?, ?>) c;
            fkCols.add(fk);
        }
    }

    // This package only uses ColumnImpl objects so we're good
    @SuppressWarnings("unchecked")
    private void setTable(Column<M, ?> c, int index) {
        assert c instanceof ColumnImpl;
        ColumnImpl<M, ?> impl = (ColumnImpl<M, ?>) c;
        impl.setTable(this);
    }


    @Override
    public Factory<M> getFactory() {
        return factory;
    }
    @Override
    public Column<M, Long> getIdColumn() {
        return idColumn;
    }
    @Override
    public Column<M, Long> getCreateTimeColumn() {
        return createTimeColumn;
    }
    @Override
    public Column<M, Long> getModifyTimeColumn() {
        return modifyTimeColumn;
    }
    @Override
    public String name() {
        return name;
    }
    @Override
    public List<Column<M, ?>> columns() {
        return columns;
    }
    @Override
    public List<Column.Fk<M, ?, ?>> fkColumns() {
        return fkColumns;
    }
    @Override
    public Column<M, ?> getNaturalKeyColumn() {
        return naturalKeyColumn;
    }
    @Override
    public Map<String, Column<M, ?>> columnsByName() {
        return columnsByName;
    }
    @Override
    public List<Column<M, ?>> getSecondaryKeyCols() {
        return secondaryKeyCols;
    }
    @Override
    public Column<M, ?> columnForName(String name) {
        return columnsByName.getOrDefault(name, null);
    }
}

File: src/main/kotlin/com/machfour/macros/core/BaseTable.kt
package com.machfour.macros.core

import java.util.Collections
import kotlin.collections.ArrayList
import kotlin.collections.LinkedHashMap

abstract class BaseTable<M>(
    final override val name: String,
    final override val factory: Factory<M>,
    cols: List<Column<M, *>>
) : Table<M> {
    final override val columns: List<Column<M, *>>

    // TODO make these better
    final override val idColumn: Column<M, Long> = cols[0] as Column<M, Long>
    final override val createTimeColumn: Column<M, Long> = cols[1] as Column<M, Long>
    final override val modifyTimeColumn: Column<M, Long> = cols[2] as Column<M, Long>

    final override val fkColumns: List<Column.Fk<M, *, *>>
    final override val columnsByName: Map<String, Column<M, *>>
    final override val secondaryKeyCols: List<Column<M, *>>
    final override val naturalKeyColumn: Column<M, *>?


    // first three columns must be ID, create time, modify time
    init {
        require(Schema.ID_COLUMN_NAME == idColumn.sqlName)
        require(Schema.CREATE_TIME_COLUMN_NAME == createTimeColumn.sqlName)
        require(Schema.MODIFY_TIME_COLUMN_NAME == modifyTimeColumn.sqlName)

        // make name map and secondary key cols list. Linked hash map keeps insertion order
        val tmpColumnsByName = LinkedHashMap<String, Column<M, *>>(cols.size, 1.0f)
        var naturalKeyColumn: Column<M, *>? = null

        val tmpSecondaryKeyCols = ArrayList<Column<M, *>>(2)
        val tmpFkColumns = ArrayList<Column.Fk<M, *, *>>(2)

        for (c in cols) {
            // This package only uses ColumnImpl objects so we're good
            require(c is ColumnImpl<*, *>)
            (c as ColumnImpl<M, *>).table = this

            tmpColumnsByName[c.sqlName] = c
            if (c.isInSecondaryKey) {
                tmpSecondaryKeyCols.add(c)
            }
            // record secondary key
            if (c.isUnique && c != idColumn) {
                require(naturalKeyColumn == null) { "two natural keys defined" }
                naturalKeyColumn = c
            }
            if (c is Column.Fk<*, *, *>) {
                tmpFkColumns.add(c as Column.Fk<M, *, *>)
            }
        }
        this.columns = Collections.unmodifiableList(cols)
        this.columnsByName = Collections.unmodifiableMap(tmpColumnsByName)
        this.secondaryKeyCols = Collections.unmodifiableList(tmpSecondaryKeyCols)
        this.fkColumns = Collections.unmodifiableList(tmpFkColumns)
        this.naturalKeyColumn = naturalKeyColumn
    }


}